platform/ios/Integration Tests/MGLShapeSourceTests.m

```diff
@@ -122,5 +122,42 @@ - (void)testSettingShapeSourceToNilInRegionIsChanging {
     [self waitForExpectations:@[expectation] timeout:1.0];
 }
 
+- (void)testShapeSourceWithLineDistanceMetrics {
+    CLLocationCoordinate2D coordinates[] = {
+        CLLocationCoordinate2DMake(9.6315313, 52.4133574),
+        CLLocationCoordinate2DMake(24.9410248, 60.1733244)};
+
+    MGLPolylineFeature *polylineFeature = [MGLPolylineFeature polylineWithCoordinates:coordinates count:sizeof(coordinates)/sizeof(coordinates[0])];
+    NSDictionary *options = @{MGLShapeSourceOptionLineDistanceMetrics: @YES};
+    MGLShapeSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"route" shape:polylineFeature options:options];
+    MGLLineStyleLayer *lineLayer = [[MGLLineStyleLayer alloc] initWithIdentifier:@"lineLayer" source:source];
+
+    [self.style addSource:source];
+    [self.style addLayer:lineLayer];
+    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(9.6315313, 52.4133574) animated:YES];
+
+    XCTestExpectation *expectation = [self expectationWithDescription:@"regionDidChange expectation"];
+    expectation.expectedFulfillmentCount = 1;
+    expectation.assertForOverFulfill = YES;
+
+    __weak id weakself = self;
+    self.regionDidChange = ^(MGLMapView *mapView, MGLCameraChangeReason reason, BOOL animated) {
+
+        id strongSelf = weakself;
+        if (!strongSelf)
+            return;
+
+        NSArray *features = [source featuresMatchingPredicate:nil];
+        MGLTestAssert(strongSelf, features.count == 1UL, @"Should contain one Feature");
+
+        MGLPolylineFeature *feature = [features objectAtIndex:0];
+        MGLTestAssertNotNil(strongSelf, [feature.attributes objectForKey:@"mapbox_clip_start"], @"Attributes should contain mapbox_clip_start property");
+        MGLTestAssertNotNil(strongSelf, [feature.attributes objectForKey:@"mapbox_clip_end"], @"Attributes should contain mapbox_clip_end property");
+
+        [expectation fulfill];
+    };
+
+    [self waitForExpectations:@[expectation] timeout:5.0];
```

Comment:
Five seconds is a long timeout for an automated test. Is it possible to replace it with a unit test of just the affected functionality? Or does the crash only occur when adding the source to the style?

Reply:
Conversion code is private to MGLFeature, I couldn't find any other place where private methods are tested , so, thought it would be better to create integration test. I can create simple unit test that verifies that exception is not throw, do you think it would be better?

Reply:
We test some private functions relating to runtime styling. The key is to declare the function with `MGL_EXPORT` in a private header and make the header private (not project visibility).

In my opinion, a unit test would be more beneficial in this case, if it’s feasible, but otherwise an integration test is fine. The only thing that caught my attention was the five-second wait time, which seems generous if we’re using a trivial style like one-liner.json.

Reply:
Unit tests seem ideal here, but we could always do that as a follow-on PR.